lib/streamlit/scriptrunner/script_runner.py

```diff
@@ -571,7 +574,14 @@ def _run_script(self, rerun_data: RerunData) -> None:
         _log_if_error(_clean_problem_modules)
 
         if rerun_exception_data is not None:
+            self.on_event.send(self, event=ScriptRunnerEvent.SCRIPT_STOPPED_FOR_RERUN)
             self._run_script(rerun_exception_data)
+        else:
+            # Signal that the script has finished. (We use SCRIPT_STOPPED_WITH_SUCCESS
+            # even if we were stopped with an exception.)
+            self.on_event.send(
+                self, event=ScriptRunnerEvent.SCRIPT_STOPPED_WITH_SUCCESS
+            )
```

Comment:
I don't have strong feelings about whether we should retain this separate `_on_script_finished` function or not - but if we're going to keep it, then I think it should be handling this event dispatching logic (it could just take the appropriate SCRIPT_STOPPED event type as a param. Otherwise we have some script-finished logic running in one place, and other logic running elsewhere.